### PR TITLE
View collection of articles based on Location

### PR DESCRIPTION
**Pull request by gsim1903**

### CODE REVIEW

**Comment by gsim1903** [comment]
Screenshot 2022-04-20 at 09 59 41

File: .env [hunk 1]
```diff
@@ -0,0 +1 @@
+REACT_APP_STRIPE_PK_KEY=pk_test_51IsORYEBWtL1DLKvi5Mf34FPtHM8Kucxy7dNDJotUW79cx9IhvERhxWCFExa1jmV8PYFi1YsYj4gCo2g72YdKCRN00hIkHfFeV
```
Comment (maxarvid) [inline .env:1]: This is out of scope for this PR, save it for the Stripe PR.

File: cypress/integration/authenticatedUserCanSeeSingleArticle.feature.js [hunk 2]
```diff
@@ -1,11 +1,30 @@
-/* eslint-disable no-undef */
 describe("Visitor  can ", () => {
   describe("see a full single article when authenticated", () => {
     beforeEach(() => {
       cy.intercept("GET", "api/articles", {
         fixture: "articles.json",
       }).as("getArticles");
```
Comment (maxarvid) [inline cypress/integration/authenticatedUserCanSeeSingleArticle.feature.js]: I think you are missing an `cy.intercept` here for the open cage api. I suggest adding the following, running your tests and make sure it is used:
```js
cy.intercept("GET", "https://api.opencagedata.com/geocode/v1/json**", {
  fixture: "location.json",
}).as("getLocation");
```

File: cypress/integration/authenticatedUserCanSeeSingleArticle.feature.js [hunk 3]
```diff
@@ -49,7 +68,22 @@ describe("Visitor  can ", () => {
         fixture: "articleShow.json",
       }).as("getSingleArticle");
```
Comment (maxarvid) [inline cypress/integration/authenticatedUserCanSeeSingleArticle.feature.js:74]: See comment above on missing `cy.intercept` for the open cage api

File: cypress/integration/userCanAccessArticleByAuthenticating.feature.js [hunk 4]
```diff
@@ -14,7 +14,16 @@ describe("visitor can see an single article by authenticating right away", () =>
...
       fixture: "authenticationSuccess.json",
     });
-    cy.visit("/");
+    cy.visit("/", {
+      onBeforeLoad(window) {
+        const response = { error: { PERMISSION_DENIED: true } };
+        cy.stub(window.navigator.geolocation, "getCurrentPosition").callsFake(
+          (callback) => {
+            return callback(response);
+          }
```
Comment (maxarvid) [inline cypress/integration/userCanAccessArticleByAuthenticating.feature.js:24]: Our TDD is made much stronger by including this scenario of the user denying location sharing, however, I'm not clear on why it should occur in this test. I can of course be convinced, but it would appear to me that it's covered by `userCanAccessArticlesBasedOnLocation.feature.js` line 57.

File: cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
+describe("Visitor can view articles based on their location ", () => {
+  beforeEach(() => {
+    cy.intercept("GET", "api/articles", {
+      fixture: "articles.json",
+    }).as("getArticles");
+
```
Comment (maxarvid) [inline cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js:6]: See comment above about stubbing out calls to the Open Cage API.

File: cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      });
+  });
+});
+
+describe("Visitor will view international articles when no location is found", () => {
+  beforeEach(() => {
+    cy.intercept("GET", "api/articles", {
+      fixture: "articles.json",
+    });
+
```
Comment (maxarvid) [inline cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js:66]: Here we face an interesting problem. Is the application calling OpenCage even if the user denies location sharing in their browser?

File: cypress/integration/userCanSeeArticleCategories.feature.js [hunk 7]
```diff
@@ -4,7 +4,23 @@ describe("Visitor can switch to sport news category tab", () => {
       fixture: "articles.json",
     }).as("getArticles");
```
Comment (maxarvid) [inline cypress/integration/userCanSeeArticleCategories.feature.js:6]: Add the `cy.intercept` for the Open Cage API here too.

File: cypress/integration/userCanSeeCollectioOfArticle.feature.js [hunk 8]
```diff
@@ -4,32 +4,42 @@ describe('user can see a collection of articles', () => {
     cy.intercept('GET', '/api/articles', {
       fixture: 'articles.json',
     }).as('getArticles')
-
```
Comment (maxarvid) [inline cypress/integration/userCanSeeCollectioOfArticle.feature.js:7]: See comment above on the Open Cage API

File: src/.env [hunk 9]
```diff
@@ -0,0 +1,3 @@
+REACT_APP_STRIPE_PK_KEY='pk_test_51IsORYEBWtL1DLKvi5Mf34FPtHM8Kucxy7dNDJotUW79cx9IhvERhxWCFExa1jmV8PYFi1YsYj4gCo2g72YdKCRN00hIkHfFeV'
+
+//const stripeProvider = loadStripe("pk_test_51IsORYEBWtL1DLKvi5Mf34FPtHM8Kucxy7dNDJotUW79cx9IhvERhxWCFExa1jmV8PYFi1YsYj4gCo2g72YdKCRN00hIkHfFeV")
```
Comment (maxarvid) [inline src/.env]: Does this file appear twice? In either case, it is out of scope for this PR, remove for now.

File: src/.env [hunk 10]
```diff
@@ -0,0 +1,3 @@
+REACT_APP_STRIPE_PK_KEY='pk_test_51IsORYEBWtL1DLKvi5Mf34FPtHM8Kucxy7dNDJotUW79cx9IhvERhxWCFExa1jmV8PYFi1YsYj4gCo2g72YdKCRN00hIkHfFeV'
+
```
Comment (maxarvid) [inline src/.env]: If this is the correct `.env` file, consider moving out the Open Cage API key here for a marginally more secure application.

File: src/App.jsx [hunk 11]
```diff
@@ -1,28 +1,28 @@
...
-
   return (
     <>
       <NavBar />
       <Routes>
         <Route path="/" element={<Articles />} />
         <Route path="/article/:id" element={<Article toast={toast} />} />
         <Route path="/:category" element={<ArticleGroup />} />
-        <Route path="/login" element={<Login toast={toast}/>} />
+        <Route path="/login" element={<Login toast={toast} />} />
```
Comment (maxarvid) [inline src/App.jsx]: Consider refactoring this to import `toast` in the relevant components instead of passing it down as props. It'll look cleaner.

File: src/Components/ArticleGroup.jsx [hunk 12]
```diff
@@ -1,38 +1,76 @@
...
+        <ul data-cy="articles-list">{articlesList}</ul>
+      </Container>
     );
-  });
+  } else {
+    const articlesList = articles[category]
+      ?.filter((object) => object["country"] === `${country}`)
+      .map((article) => {
+        return (
+          <div key={article.id} style={{ listStyleType: "none" }}>
```
Comment (maxarvid) [inline src/Components/ArticleGroup.jsx]: Consider changing these `div`s to `li`s

File: src/Components/Articles.jsx [hunk 13]
```diff
@@ -27,8 +27,9 @@ const Articles = () => {
 
   const articleList = displayArticles(articles).map((article) => {
     return (
-      <Card>
-        <li key={article.id} style={{ listStyleType: "none" }}>
+      
+        <div key={article.id} style={{ listStyleType: "none" }}>
```
Comment (maxarvid) [inline src/Components/Articles.jsx:31]: It's not the `li` tag triggering the warning, it's that `<Card>` was wrapped around the `<li>`s and not the other way around.

File: src/Components/Location.jsx [hunk 14]
```diff
@@ -0,0 +1,49 @@
+import React, { useEffect } from 'react'
```
Comment (maxarvid) [inline src/Components/Location.jsx:1]: Format the code for consitency

File: src/Components/Location.jsx [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  }, [])
+
+  const fetchGeolocation = async () => {
+    const position = await fetchUserCoordinates()
+
+    if (position.error) {
+      dispatch({
+        type: 'SET_USER_COUNTRY',
+        payload: 'location cannot be detected',
+      })
```
Comment (maxarvid) [inline src/Components/Location.jsx:28]: This is great, the application does not call the Open Cage API if the user does not share location. Is there a test to make sure this is the case?

**Review by maxarvid — CHANGES_REQUESTED** [review]
Awesome bug hunting, made some comments mostly on a missing intercept for the Open Cage API.

File: .env (hunk 1 above)
Comment (gsim1903) [reply .env:1]: Understand. Added to git ignore so should be away from PR

File: cypress/integration/authenticatedUserCanSeeSingleArticle.feature.js (hunk 2 above)
Comment (gsim1903) [reply cypress/integration/authenticatedUserCanSeeSingleArticle.feature.js]: Screenshot 2022-04-22 at 06 39 41


Added

File: cypress/integration/authenticatedUserCanSeeSingleArticle.feature.js (hunk 3 above)
Comment (gsim1903) [reply cypress/integration/authenticatedUserCanSeeSingleArticle.feature.js:74]: Updated

File: cypress/integration/userCanAccessArticleByAuthenticating.feature.js (hunk 4 above)
Comment (gsim1903) [reply cypress/integration/userCanAccessArticleByAuthenticating.feature.js:24]: I could well be wrong, but here goes my thinking at least..

The feature the test was written for was to still show articles if no location is detected from user. Effectively to return all articles when no location present. 

UserCanAccessArticlesByAuthenticating   tested ability to view articles once signed up. But were note testing articles sorted / selected by country type which I undertstood was the other test

File: cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js (hunk 5 above)
Comment (gsim1903) [reply cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js:6]: I've added though I'm fairly sure on review of test draft Thomas suggested to remove API stubbing as code should kick in to stop that being called with the stub location
Screenshot 2022-04-22 at 06 53 22
?

File: src/Components/ArticleGroup.jsx (hunk 12 above)
Comment (gsim1903) [reply src/Components/ArticleGroup.jsx]: Screenshot 2022-04-22 at 06 55 28


updated

File: src/Components/Location.jsx (hunk 15 above)
Comment (gsim1903) [reply src/Components/Location.jsx:28]: /Users/gavinsimpson/dgms-news-end-user-client/cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js
/79

  it('is expected to show that the users location is not found', () => {
    cy.get('[data-cy=user-location]').should(
      'contain',
      'location cannot be detected',

Does this test? (have reworded since original PR)

File: src/.env (hunk 9 above)
Comment (gsim1903) [reply src/.env]: Removed

File: src/Components/Articles.jsx (hunk 13 above)
Comment (gsim1903) [reply src/Components/Articles.jsx:31]: Warning now gone. Thanks 
Screenshot 2022-04-22 at 07 16 21

File: cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js (hunk 6 above)
Comment (gsim1903) [reply cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js:66]: I thought not,  it a position error occurs from geolocation, the if statement returns a value to redux for user country, the alternative (i.e. that the user location is shared) being that the coordinates are used to return the country? Or maybe you are nicely hinting this doesn't work as it should?

Screenshot 2022-04-22 at 07 23 55

Comment (gsim1903) [reply cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js:66]: Screenshot 2022-04-22 at 07 27 27

If I add open cage stub to the sad path it seems its not called, which is what we want right?

File: cypress/integration/userCanSeeArticleCategories.feature.js (hunk 7 above)
Comment (gsim1903) [reply cypress/integration/userCanSeeArticleCategories.feature.js:6]: Added

File: cypress/integration/userCanSeeCollectioOfArticle.feature.js (hunk 8 above)
Comment (gsim1903) [reply cypress/integration/userCanSeeCollectioOfArticle.feature.js:7]: Screenshot 2022-04-22 at 07 29 28

Added

File: src/.env (hunk 10 above)
Comment (gsim1903) [reply src/.env]: Screenshot 2022-04-22 at 07 38 21


Thanks a useful revisit to get this to work!

File: src/App.jsx (hunk 11 above)
Comment (gsim1903) [reply src/App.jsx]: Screenshot 2022-04-22 at 07 53 23

Done

**Comment by gsim1903** [comment]
@maxarvid  thanks for comment, have hopefully comments and or resolved all.

File: src/Components/Location.jsx (hunk 15 above)
Comment (maxarvid) [reply src/Components/Location.jsx:28]: Yes! Good stuff.

File: cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js (hunk 6 above)
Comment (maxarvid) [reply cypress/integration/userCanAccessArticlesBasedOnLocation.feauture.js:66]: Yes, this is the correct functionality ATMO

File: src/Components/login.css [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    width: 118px;
+    margin-bottom: -10px;
+  }
+  .login-form {
+    margin-top: 35px;
+  }
+  .login-input-field {
+    font-size: 18px;
+    margin-bottom: 25px;
+  }
```
Comment (maxarvid) [inline src/Components/login.css:33]: This `.css` file should not live with the components. Consider moving it to either a new `style` directory or move it to `index.css`

**Review by maxarvid — CHANGES_REQUESTED** [review]
Excellent, one last thing about the .css file location that I missed the 1st time

File: src/Components/login.css (hunk 16 above)
Comment (maxarvid) [reply src/Components/login.css:33]: In the interest of time, we will create a chore and sort this out at a later stage.

**Review by maxarvid — APPROVED** [review]
